py/desispec/preproc.py

```diff
@@ -840,6 +853,79 @@ def read_bias(filename=None, camera=None, dateobs=None):
+    profile_2d_Right=np.transpose(np.tile(profileRight,(int(ny/2),1)))
+    profile_2d=np.concatenate((profile_2d_Left,profile_2d_Right),axis=1)
+    image=profile_2d+hdus['DARK'].data*float(exptime) # Note no zero is added
+    return image
+
+
+def read_2d_dark(filename=None, exptime=0):
+
+    '''
+    Return exptime dependent dark+2D profile for camera on dateobs or night
+
+    Options:
+        filename : input filename to read
+        exptime : the exposure time of the image
+    Notes:
+        must provide filename
+    '''
+    from astropy.io import fits
+    log=get_logger()
+    if filename is None:
+        #- use camera and dateobs to derive what bias file should be used
+        raise NotImplementedError
+    else:
+        hdus = fits.open(filename)
+        if not str(exptime).isnumeric(): # if the exptime is not a number, use exptime=0 for default
+            exptime=0
+        exptime_arr=[]
+        ext_arr={}
+        for hdu in hdus:
+            if hdu.header['EXTNAME'] == 'DARK':
+                pass
+            elif hdu.header['EXTNAME'] == 'ZERO':
+                ext_arr['0']=hdu.header['EXTNAME']
+                exptime_arr.append(0)
+            else:
+                ext_arr[hdu.header['EXTNAME'][1:]]=hdu.header['EXTNAME']
+                exptime_arr.append(int(hdu.header['EXTNAME'][1:]))
+        #log.info('ext_arr',ext_arr)
+        #log.info('exptime_arr',exptime_arr)
+        if int(exptime)==0:
+            return hdus['DARK'].data*0.# No dark, return a zero array
+        elif int(exptime) in exptime_arr:
+            log.info('Using bias+dark at exptime='+str(int(exptime)))
+            return recover_2d_dark(hdus,exptime,ext_arr[str(int(exptime))])  #hdus[str(int(exptime))].data
+        elif int(exptime)> max(exptime_arr):
+            log.info('Using bias+dark at exptime='+str(max(exptime_arr)))
+            return recover_2d_dark(hdus,max(exptime_arr),ext_arr[str(int(max(exptime_arr)))])  #hdus[str(max(exptime_arr))].data
+        elif int(exptime)< min(exptime_arr):
+            log.info('Using bias+dark at exptime='+str(min(exptime_arr)))
+            return recover_2d_dark(hdus,min(exptime_arr),ext_arr[str(int(min(exptime_arr)))])  #hdus[str(min(exptime_arr))].data
+        else:
+            # Interpolate
+            exptime_arr=np.sort(np.array(exptime_arr))
+            ind=np.where(exptime_arr>int(exptime))
+            ind1=ind[0][0]-1
+            ind2=ind[0][0]
+            log.info('Interpolate between '+str(exptime_arr[ind1])+' and '+str(exptime_arr[ind2]))
+            precision=(float(exptime)-exptime_arr[ind1])/(exptime_arr[ind2]-exptime_arr[ind1])
+            # Run interpolation
+            image1=recover_2d_bias_dark(hdus,exptime_arr[ind1],ext_arr[str(int(exptime_arr[ind1]))])
+            image2=recover_2d_bias_dark(hdus,exptime_arr[ind2],ext_arr[str(int(exptime_arr[ind2]))])
+            img=image1*precision+image2*(1-precision) #interp_shape(image1,image2,precision)
```

Comment:
Isn't it the opposite? because here precision=0 if exptime=exptime_arr[ind1].

Reply:
You are absolutely right